temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+                activity.worker_shutdown_event.set()
+        if activity_tasks:
+            # Wait for any still running after graceful timeout
+            _, still_running = await asyncio.wait(
+                activity_tasks, timeout=graceful_timeout.total_seconds()
+            )
+            if still_running:
+                # Cancel all still running
+                if len(still_running) == 1:
+                    logger.info(f"Cancelling 1 activity that is still running")
+                else:
+                    logger.info(
+                        f"Cancelling {len(still_running)} activities that are still running"
+                    )
+                for task in still_running:
+                    # We have to find the running activity that's associated with
+                    # the task so that we can cancel through that. It's ok if the
+                    # activity is already gone.
+                    for activity in self._running_activities.values():
+                        if activity.task is task:
+                            activity.cancel()
+                            break
+
+        # Wait for the bridge to report all activities are completed
+        await bridge_shutdown_task
+
+    async def _run_activities(self) -> None:
+        # Continually poll for activity work
+        while True:
+            try:
+                # Poll for a task
+                task = await self._bridge_worker.poll_activity_task()
+
+                if task.HasField("start"):
+                    # Cancelled event and sync field will be updated inside
+                    # _run_activity when the activity function is obtained
+                    activity = _RunningActivity(
+                        task=None,
+                        cancelled_event=temporalio.activity._CompositeEvent(
+                            thread_event=threading.Event(), async_event=None
+                        ),
+                        worker_shutdown_event=temporalio.activity._CompositeEvent(
+                            thread_event=threading.Event(), async_event=None
+                        ),
+                        sync=False,
+                        last_heartbeat_task=None,
+                    )
+                    activity.task = asyncio.create_task(
+                        self._run_activity(task.task_token, task.start, activity)
+                    )
+                    self._running_activities[task.task_token] = activity
+                elif task.HasField("cancel"):
+                    self._cancel_activity(task.task_token, task.cancel)
+                else:
+                    logger.warning("unrecognized activity task: %s", task)
+            except temporalio.bridge.worker.PollShutdownError:
+                return
+            except Exception as err:
+                # Should never happen
+                logger.exception(f"Activity runner failed: {err}", extra={"error": err})
```

Comment:
logger.exception should already include the error message and stack trace

Reply:
You generally don't put errors in python logs, `exc_info` does that automatically.
`exception` is just an alias to `error(exc_info=True)`

Reply:
Please remove because the way it is now errors are displayed twice

Reply:
Will do

Reply:
Done.